app/lib/account/consent_backend.dart

```diff
@@ -0,0 +1,176 @@
+      }
+      if (result.granted ?? false) {
+        await _accept(c);
+        return api.ConsentResult(granted: true);
+      } else {
+        await _delete(c);
+        return api.ConsentResult(granted: false);
+      }
+    });
+  }
+
+  /// Create a new invitation, or
+  /// - if it already exists, re-send the notification, or
+  /// - if it was sent recently, do nothing.
+  Future<InviteStatus> invite({
+    @required String userId,
+    @required String type,
+    @required List<String> args,
+    @required String descriptionText,
+    @required String descriptionHtml,
+  }) async {
+    return retry(() async {
+      return await withAuthenticatedUser((activeUser) async {
+        // First check for existing consents with identical dedupId.
+        final dedupId = consentDedupId(type, args);
+        final userKey = _db.emptyKey.append(User, id: userId);
+        final query = _db.query<Consent>(ancestorKey: userKey)
+          ..filter('dedupId =', dedupId);
+        final list = await query.run().toList();
+        if (list.isNotEmpty) {
+          final old = list.first;
+          if (old.isExpired()) {
+            // expired entries should be deleted
+            await _delete(old);
+          } else if (old.shouldNotify()) {
+            // non-expired entries just re-send the notification
+            return await _sendNotification(old);
+          } else {
+            return InviteStatus(false, old.nextNotification);
+          }
+        }
+        // Create a new entry.
+        final consent = Consent.init(
+          parentKey: userKey,
+          type: type,
+          args: args,
+          fromUserId: authenticatedUser.userId,
+          descriptionText: descriptionText,
+          descriptionHtml: descriptionHtml,
+        );
+        await _db.commit(inserts: [consent]);
+        return await _sendNotification(consent);
+      });
+    });
+  }
+
+  Future<InviteStatus> _sendNotification(Consent consent) async {
+    final invitedEmail = await accountBackend.getEmailOfUserId(consent.userId);
+    await emailSender.sendMessage(createInviteEmail(
+      activeAccountEmail: authenticatedUser.email,
+      invitedEmail: invitedEmail,
+      message: consent.descriptionText,
+      consentUrl: consentUrl(consent.consentId),
+    ));
+    return await _db.withTransaction((tx) async {
+      final c = (await tx.lookup<Consent>([consent.key])).single;
+      c.notificationCount++;
+      c.lastNotified = DateTime.now().toUtc();
+      tx.queueMutations(inserts: [c]);
+      await tx.commit();
+      return InviteStatus(true, c.nextNotification);
+    }) as InviteStatus;
+  }
+
+  /// Removes obsolete/expired [Consent] entries from Datastore.
+  Future deleteObsoleteConsents() async {
+    final query = _db.query<Consent>()
+      ..filter('expires <', DateTime.now().toUtc());
+    await for (var entry in query.run()) {
+      try {
+        await _delete(entry);
+      } catch (e) {
+        _logger.info(
```

Comment:
please make it a warning. This is problematic... Also please handle exceptions in the case where the entity doesn't exist, because of racing delete requests.

Reply:
IMO, we should handle everything that could go wrong, and leave the rest as `_logger.shout`.